Assignment3/Q1.py

```diff
@@ -0,0 +1,128 @@
+            columns: an integer, number of columns in a grid
+            grid: a 2D array of characters, we are searching for words from dictionary in a grid
+            dictionary: a Dictionary object, contains the words to search for in a grid
+
+        Returns:
+            a set of words, complete list of words that can be formed from grid's characters and are in a dictionary,
+            there are no duplicated words in a set.
+    """
+    # Instantiate a list of words.
+    word_list = []
+
+    # Create a graph corresponding to the given grid.
+    graph = create_corresponding_graph(rows, columns)
+
+    # For each vertex in a graph perform a search from vertex to find all the words starting with corresponding
+    # character in a grid.
+    for key in graph:
+        words = word_search_recursive(dictionary, grid, graph, key)
+        word_list = word_list + words
+
+    # Remove duplicates from the word list.
+    word_list = set(word_list)
+
+    # Return a set of words found in a grid and contained in a dictionary.
+    return word_list
+
+
+def word_search_recursive(dictionary, grid, graph, start, prefix="", words=[], path=[]):
+    """This method works similar to DFS from vertex on a graph but here it takes ALL paths from the vertex
+       to find all the words that can be formed from a grid and are contained in a dictionary.
+
+        Args:
+            dictionary: a Dictionary object, contains the words to search for in a grid
+            grid: 2D array of characters, we are searching for words from dictionary in a grid
+            graph: a adjacency list (using Python dictionary) representation of a graph corresponding
+                   to the given grid. Each vertes in a graph is a tuple of coordinates of a cell in a grid
+            start: an integer tuple, the vertex we start our search from
+            prefix: a string, contains a word 'accumulated' so far, in a recursive call prefix is modify in order
+                    to perform a DFS search, we terminate early if isPrefix(string) on a current prefix returns False
+            words: a list of strings, list of words that can be formulated from a character grid and are
+                   in a dictionary
+            path: list of integer tuples (similar to path in DFS search) we keep here an information about the vertices
+                  visited so far in a search, it prevents visiting the same cell in a grid twice in a single word search
+
+        Returns:
+            a list of strings, list of words that can be form using characters in a grid, stating from start vertex
+            (corresponding to some character in a grid) that are in a dictionary.
+    """
+    # Get coordinates of the starting vertex in a search.
+    (i, j) = start
+
+    # Add a corresponding character to the prefix 'accumulated' so far.
+    # NOTE: We are case insensitive in this solution
+    prefix = prefix + grid[i][j].upper()
```

Comment:
As noted elsewhere, we now have two separate pieces of code (the implementation of this method and of the dictionary) which both need to know that the convention is to convert everything to upper case.